pkg/webhook/pod/mutating/device_resource_spec.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutating
+
+import (
+	"context"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+)
+
+func (h *PodMutatingHandler) deviceResourceSpecMutatingPod(ctx context.Context, req admission.Request, pod *corev1.Pod) error {
+	if req.Operation != admissionv1.Create && req.Operation != admissionv1.Update {
+		return nil
+	}
+
+	return h.mutateByDeviceResources(pod)
+}
+
+func (h *PodMutatingHandler) mutateByDeviceResources(pod *corev1.Pod) error {
+	// device resource request euqal limit, not overcommit
+	for i := range pod.Spec.Containers {
+		container := &pod.Spec.Containers[i]
+		r := getContainerExtendedResourcesRequirement(container, []corev1.ResourceName{
+			extension.ResourceGPU,
+			extension.ResourceGPUCore,
+			extension.ResourceGPUMemory,
+			extension.ResourceGPUMemoryRatio,
+			extension.ResourceGPUShared,
+		})
+		if r == nil {
+			continue
+		}
+
+		// use gpu api first
+		_, ok := r.Requests[extension.ResourceGPU]
+		if ok {
+			injectGPU(container)
+			continue
```

Comment:
why continue here, koordinator.sh/gpu quantiy just mean koordinator.sh/gpu-core and koordinator.sh/gpu-memory-ratio both
equals it, we should continue to process gpu.shared logic

Reply:
convert koordinator.sh/gpu to koordinator.sh/gpu-core and koordinator.sh/gpu-memory-ratio, than process gpu.shared